# Scraper for "Enti e servizi presenti su IO"

## Introductions

[iO](https://io.italia.it/) is an app developed by the Italian Ministry of Technological Innovation with the goal of creating a a single channel for institutions and citizens where it's possible to access all digital public services.

The goal of this project is to [obtain a list of all institutions](https://io.italia.it/enti/) that are using iO for their services.

In [1]:
# libraries for requesting url and scraping jsons
import requests
import json

In [2]:
# getting unique IDs for all institutions
request = requests.get('https://io.italia.it/assets/json/enti-list-searchable.json')
file = request.json()

codici_ipa = []
for element in file:
    ente = i.split('|')
    codici_ipa.append(ente[1])

In [3]:
print(codici_ipa[0:5])

['00404190787', '00421140039', '01447330190', '83002690275', '80228830586']


In [4]:
len(codici_ipa) 

7228

We can see that there are 7228 unique IDs, ranging from small public administrations like "Comuni" to Regions, Metropolitan Cities and government offices.

In [6]:
count = 0
data = []
for codice in codici_ipa:
    website = 'https://io.italia.it/assets/entijson/'+codice+'.json'
    cont = requests.get(website)
    parsed = cont.json()
    count = count + 1
    print("I have scraped "+str(count)+" objects.")
    data.append(parsed)
    
data_to_save = open("data.json", "w")
json = json.dump(data, data_to_save)
data_to_save.close()

    

I have scraped 1 objects.
I have scraped 2 objects.
I have scraped 3 objects.
I have scraped 4 objects.
I have scraped 5 objects.
I have scraped 6 objects.
I have scraped 7 objects.
I have scraped 8 objects.
I have scraped 9 objects.
I have scraped 10 objects.
I have scraped 11 objects.
I have scraped 12 objects.
I have scraped 13 objects.
I have scraped 14 objects.
I have scraped 15 objects.
I have scraped 16 objects.
I have scraped 17 objects.
I have scraped 18 objects.
I have scraped 19 objects.
I have scraped 20 objects.
I have scraped 21 objects.
I have scraped 22 objects.
I have scraped 23 objects.
I have scraped 24 objects.
I have scraped 25 objects.
I have scraped 26 objects.
I have scraped 27 objects.
I have scraped 28 objects.
I have scraped 29 objects.
I have scraped 30 objects.
I have scraped 31 objects.
I have scraped 32 objects.
I have scraped 33 objects.
I have scraped 34 objects.
I have scraped 35 objects.
I have scraped 36 objects.
I have scraped 37 objects.
I have scr

I have scraped 299 objects.
I have scraped 300 objects.
I have scraped 301 objects.
I have scraped 302 objects.
I have scraped 303 objects.
I have scraped 304 objects.
I have scraped 305 objects.
I have scraped 306 objects.
I have scraped 307 objects.
I have scraped 308 objects.
I have scraped 309 objects.
I have scraped 310 objects.
I have scraped 311 objects.
I have scraped 312 objects.
I have scraped 313 objects.
I have scraped 314 objects.
I have scraped 315 objects.
I have scraped 316 objects.
I have scraped 317 objects.
I have scraped 318 objects.
I have scraped 319 objects.
I have scraped 320 objects.
I have scraped 321 objects.
I have scraped 322 objects.
I have scraped 323 objects.
I have scraped 324 objects.
I have scraped 325 objects.
I have scraped 326 objects.
I have scraped 327 objects.
I have scraped 328 objects.
I have scraped 329 objects.
I have scraped 330 objects.
I have scraped 331 objects.
I have scraped 332 objects.
I have scraped 333 objects.
I have scraped 334 o

I have scraped 593 objects.
I have scraped 594 objects.
I have scraped 595 objects.
I have scraped 596 objects.
I have scraped 597 objects.
I have scraped 598 objects.
I have scraped 599 objects.
I have scraped 600 objects.
I have scraped 601 objects.
I have scraped 602 objects.
I have scraped 603 objects.
I have scraped 604 objects.
I have scraped 605 objects.
I have scraped 606 objects.
I have scraped 607 objects.
I have scraped 608 objects.
I have scraped 609 objects.
I have scraped 610 objects.
I have scraped 611 objects.
I have scraped 612 objects.
I have scraped 613 objects.
I have scraped 614 objects.
I have scraped 615 objects.
I have scraped 616 objects.
I have scraped 617 objects.
I have scraped 618 objects.
I have scraped 619 objects.
I have scraped 620 objects.
I have scraped 621 objects.
I have scraped 622 objects.
I have scraped 623 objects.
I have scraped 624 objects.
I have scraped 625 objects.
I have scraped 626 objects.
I have scraped 627 objects.
I have scraped 628 o

I have scraped 887 objects.
I have scraped 888 objects.
I have scraped 889 objects.
I have scraped 890 objects.
I have scraped 891 objects.
I have scraped 892 objects.
I have scraped 893 objects.
I have scraped 894 objects.
I have scraped 895 objects.
I have scraped 896 objects.
I have scraped 897 objects.
I have scraped 898 objects.
I have scraped 899 objects.
I have scraped 900 objects.
I have scraped 901 objects.
I have scraped 902 objects.
I have scraped 903 objects.
I have scraped 904 objects.
I have scraped 905 objects.
I have scraped 906 objects.
I have scraped 907 objects.
I have scraped 908 objects.
I have scraped 909 objects.
I have scraped 910 objects.
I have scraped 911 objects.
I have scraped 912 objects.
I have scraped 913 objects.
I have scraped 914 objects.
I have scraped 915 objects.
I have scraped 916 objects.
I have scraped 917 objects.
I have scraped 918 objects.
I have scraped 919 objects.
I have scraped 920 objects.
I have scraped 921 objects.
I have scraped 922 o

I have scraped 1175 objects.
I have scraped 1176 objects.
I have scraped 1177 objects.
I have scraped 1178 objects.
I have scraped 1179 objects.
I have scraped 1180 objects.
I have scraped 1181 objects.
I have scraped 1182 objects.
I have scraped 1183 objects.
I have scraped 1184 objects.
I have scraped 1185 objects.
I have scraped 1186 objects.
I have scraped 1187 objects.
I have scraped 1188 objects.
I have scraped 1189 objects.
I have scraped 1190 objects.
I have scraped 1191 objects.
I have scraped 1192 objects.
I have scraped 1193 objects.
I have scraped 1194 objects.
I have scraped 1195 objects.
I have scraped 1196 objects.
I have scraped 1197 objects.
I have scraped 1198 objects.
I have scraped 1199 objects.
I have scraped 1200 objects.
I have scraped 1201 objects.
I have scraped 1202 objects.
I have scraped 1203 objects.
I have scraped 1204 objects.
I have scraped 1205 objects.
I have scraped 1206 objects.
I have scraped 1207 objects.
I have scraped 1208 objects.
I have scraped

I have scraped 1458 objects.
I have scraped 1459 objects.
I have scraped 1460 objects.
I have scraped 1461 objects.
I have scraped 1462 objects.
I have scraped 1463 objects.
I have scraped 1464 objects.
I have scraped 1465 objects.
I have scraped 1466 objects.
I have scraped 1467 objects.
I have scraped 1468 objects.
I have scraped 1469 objects.
I have scraped 1470 objects.
I have scraped 1471 objects.
I have scraped 1472 objects.
I have scraped 1473 objects.
I have scraped 1474 objects.
I have scraped 1475 objects.
I have scraped 1476 objects.
I have scraped 1477 objects.
I have scraped 1478 objects.
I have scraped 1479 objects.
I have scraped 1480 objects.
I have scraped 1481 objects.
I have scraped 1482 objects.
I have scraped 1483 objects.
I have scraped 1484 objects.
I have scraped 1485 objects.
I have scraped 1486 objects.
I have scraped 1487 objects.
I have scraped 1488 objects.
I have scraped 1489 objects.
I have scraped 1490 objects.
I have scraped 1491 objects.
I have scraped

I have scraped 1741 objects.
I have scraped 1742 objects.
I have scraped 1743 objects.
I have scraped 1744 objects.
I have scraped 1745 objects.
I have scraped 1746 objects.
I have scraped 1747 objects.
I have scraped 1748 objects.
I have scraped 1749 objects.
I have scraped 1750 objects.
I have scraped 1751 objects.
I have scraped 1752 objects.
I have scraped 1753 objects.
I have scraped 1754 objects.
I have scraped 1755 objects.
I have scraped 1756 objects.
I have scraped 1757 objects.
I have scraped 1758 objects.
I have scraped 1759 objects.
I have scraped 1760 objects.
I have scraped 1761 objects.
I have scraped 1762 objects.
I have scraped 1763 objects.
I have scraped 1764 objects.
I have scraped 1765 objects.
I have scraped 1766 objects.
I have scraped 1767 objects.
I have scraped 1768 objects.
I have scraped 1769 objects.
I have scraped 1770 objects.
I have scraped 1771 objects.
I have scraped 1772 objects.
I have scraped 1773 objects.
I have scraped 1774 objects.
I have scraped

I have scraped 2024 objects.
I have scraped 2025 objects.
I have scraped 2026 objects.
I have scraped 2027 objects.
I have scraped 2028 objects.
I have scraped 2029 objects.
I have scraped 2030 objects.
I have scraped 2031 objects.
I have scraped 2032 objects.
I have scraped 2033 objects.
I have scraped 2034 objects.
I have scraped 2035 objects.
I have scraped 2036 objects.
I have scraped 2037 objects.
I have scraped 2038 objects.
I have scraped 2039 objects.
I have scraped 2040 objects.
I have scraped 2041 objects.
I have scraped 2042 objects.
I have scraped 2043 objects.
I have scraped 2044 objects.
I have scraped 2045 objects.
I have scraped 2046 objects.
I have scraped 2047 objects.
I have scraped 2048 objects.
I have scraped 2049 objects.
I have scraped 2050 objects.
I have scraped 2051 objects.
I have scraped 2052 objects.
I have scraped 2053 objects.
I have scraped 2054 objects.
I have scraped 2055 objects.
I have scraped 2056 objects.
I have scraped 2057 objects.
I have scraped

I have scraped 2308 objects.
I have scraped 2309 objects.
I have scraped 2310 objects.
I have scraped 2311 objects.
I have scraped 2312 objects.
I have scraped 2313 objects.
I have scraped 2314 objects.
I have scraped 2315 objects.
I have scraped 2316 objects.
I have scraped 2317 objects.
I have scraped 2318 objects.
I have scraped 2319 objects.
I have scraped 2320 objects.
I have scraped 2321 objects.
I have scraped 2322 objects.
I have scraped 2323 objects.
I have scraped 2324 objects.
I have scraped 2325 objects.
I have scraped 2326 objects.
I have scraped 2327 objects.
I have scraped 2328 objects.
I have scraped 2329 objects.
I have scraped 2330 objects.
I have scraped 2331 objects.
I have scraped 2332 objects.
I have scraped 2333 objects.
I have scraped 2334 objects.
I have scraped 2335 objects.
I have scraped 2336 objects.
I have scraped 2337 objects.
I have scraped 2338 objects.
I have scraped 2339 objects.
I have scraped 2340 objects.
I have scraped 2341 objects.
I have scraped

I have scraped 2592 objects.
I have scraped 2593 objects.
I have scraped 2594 objects.
I have scraped 2595 objects.
I have scraped 2596 objects.
I have scraped 2597 objects.
I have scraped 2598 objects.
I have scraped 2599 objects.
I have scraped 2600 objects.
I have scraped 2601 objects.
I have scraped 2602 objects.
I have scraped 2603 objects.
I have scraped 2604 objects.
I have scraped 2605 objects.
I have scraped 2606 objects.
I have scraped 2607 objects.
I have scraped 2608 objects.
I have scraped 2609 objects.
I have scraped 2610 objects.
I have scraped 2611 objects.
I have scraped 2612 objects.
I have scraped 2613 objects.
I have scraped 2614 objects.
I have scraped 2615 objects.
I have scraped 2616 objects.
I have scraped 2617 objects.
I have scraped 2618 objects.
I have scraped 2619 objects.
I have scraped 2620 objects.
I have scraped 2621 objects.
I have scraped 2622 objects.
I have scraped 2623 objects.
I have scraped 2624 objects.
I have scraped 2625 objects.
I have scraped

I have scraped 2876 objects.
I have scraped 2877 objects.
I have scraped 2878 objects.
I have scraped 2879 objects.
I have scraped 2880 objects.
I have scraped 2881 objects.
I have scraped 2882 objects.
I have scraped 2883 objects.
I have scraped 2884 objects.
I have scraped 2885 objects.
I have scraped 2886 objects.
I have scraped 2887 objects.
I have scraped 2888 objects.
I have scraped 2889 objects.
I have scraped 2890 objects.
I have scraped 2891 objects.
I have scraped 2892 objects.
I have scraped 2893 objects.
I have scraped 2894 objects.
I have scraped 2895 objects.
I have scraped 2896 objects.
I have scraped 2897 objects.
I have scraped 2898 objects.
I have scraped 2899 objects.
I have scraped 2900 objects.
I have scraped 2901 objects.
I have scraped 2902 objects.
I have scraped 2903 objects.
I have scraped 2904 objects.
I have scraped 2905 objects.
I have scraped 2906 objects.
I have scraped 2907 objects.
I have scraped 2908 objects.
I have scraped 2909 objects.
I have scraped

I have scraped 3160 objects.
I have scraped 3161 objects.
I have scraped 3162 objects.
I have scraped 3163 objects.
I have scraped 3164 objects.
I have scraped 3165 objects.
I have scraped 3166 objects.
I have scraped 3167 objects.
I have scraped 3168 objects.
I have scraped 3169 objects.
I have scraped 3170 objects.
I have scraped 3171 objects.
I have scraped 3172 objects.
I have scraped 3173 objects.
I have scraped 3174 objects.
I have scraped 3175 objects.
I have scraped 3176 objects.
I have scraped 3177 objects.
I have scraped 3178 objects.
I have scraped 3179 objects.
I have scraped 3180 objects.
I have scraped 3181 objects.
I have scraped 3182 objects.
I have scraped 3183 objects.
I have scraped 3184 objects.
I have scraped 3185 objects.
I have scraped 3186 objects.
I have scraped 3187 objects.
I have scraped 3188 objects.
I have scraped 3189 objects.
I have scraped 3190 objects.
I have scraped 3191 objects.
I have scraped 3192 objects.
I have scraped 3193 objects.
I have scraped

I have scraped 3445 objects.
I have scraped 3446 objects.
I have scraped 3447 objects.
I have scraped 3448 objects.
I have scraped 3449 objects.
I have scraped 3450 objects.
I have scraped 3451 objects.
I have scraped 3452 objects.
I have scraped 3453 objects.
I have scraped 3454 objects.
I have scraped 3455 objects.
I have scraped 3456 objects.
I have scraped 3457 objects.
I have scraped 3458 objects.
I have scraped 3459 objects.
I have scraped 3460 objects.
I have scraped 3461 objects.
I have scraped 3462 objects.
I have scraped 3463 objects.
I have scraped 3464 objects.
I have scraped 3465 objects.
I have scraped 3466 objects.
I have scraped 3467 objects.
I have scraped 3468 objects.
I have scraped 3469 objects.
I have scraped 3470 objects.
I have scraped 3471 objects.
I have scraped 3472 objects.
I have scraped 3473 objects.
I have scraped 3474 objects.
I have scraped 3475 objects.
I have scraped 3476 objects.
I have scraped 3477 objects.
I have scraped 3478 objects.
I have scraped

I have scraped 3729 objects.
I have scraped 3730 objects.
I have scraped 3731 objects.
I have scraped 3732 objects.
I have scraped 3733 objects.
I have scraped 3734 objects.
I have scraped 3735 objects.
I have scraped 3736 objects.
I have scraped 3737 objects.
I have scraped 3738 objects.
I have scraped 3739 objects.
I have scraped 3740 objects.
I have scraped 3741 objects.
I have scraped 3742 objects.
I have scraped 3743 objects.
I have scraped 3744 objects.
I have scraped 3745 objects.
I have scraped 3746 objects.
I have scraped 3747 objects.
I have scraped 3748 objects.
I have scraped 3749 objects.
I have scraped 3750 objects.
I have scraped 3751 objects.
I have scraped 3752 objects.
I have scraped 3753 objects.
I have scraped 3754 objects.
I have scraped 3755 objects.
I have scraped 3756 objects.
I have scraped 3757 objects.
I have scraped 3758 objects.
I have scraped 3759 objects.
I have scraped 3760 objects.
I have scraped 3761 objects.
I have scraped 3762 objects.
I have scraped

I have scraped 4012 objects.
I have scraped 4013 objects.
I have scraped 4014 objects.
I have scraped 4015 objects.
I have scraped 4016 objects.
I have scraped 4017 objects.
I have scraped 4018 objects.
I have scraped 4019 objects.
I have scraped 4020 objects.
I have scraped 4021 objects.
I have scraped 4022 objects.
I have scraped 4023 objects.
I have scraped 4024 objects.
I have scraped 4025 objects.
I have scraped 4026 objects.
I have scraped 4027 objects.
I have scraped 4028 objects.
I have scraped 4029 objects.
I have scraped 4030 objects.
I have scraped 4031 objects.
I have scraped 4032 objects.
I have scraped 4033 objects.
I have scraped 4034 objects.
I have scraped 4035 objects.
I have scraped 4036 objects.
I have scraped 4037 objects.
I have scraped 4038 objects.
I have scraped 4039 objects.
I have scraped 4040 objects.
I have scraped 4041 objects.
I have scraped 4042 objects.
I have scraped 4043 objects.
I have scraped 4044 objects.
I have scraped 4045 objects.
I have scraped

I have scraped 4295 objects.
I have scraped 4296 objects.
I have scraped 4297 objects.
I have scraped 4298 objects.
I have scraped 4299 objects.
I have scraped 4300 objects.
I have scraped 4301 objects.
I have scraped 4302 objects.
I have scraped 4303 objects.
I have scraped 4304 objects.
I have scraped 4305 objects.
I have scraped 4306 objects.
I have scraped 4307 objects.
I have scraped 4308 objects.
I have scraped 4309 objects.
I have scraped 4310 objects.
I have scraped 4311 objects.
I have scraped 4312 objects.
I have scraped 4313 objects.
I have scraped 4314 objects.
I have scraped 4315 objects.
I have scraped 4316 objects.
I have scraped 4317 objects.
I have scraped 4318 objects.
I have scraped 4319 objects.
I have scraped 4320 objects.
I have scraped 4321 objects.
I have scraped 4322 objects.
I have scraped 4323 objects.
I have scraped 4324 objects.
I have scraped 4325 objects.
I have scraped 4326 objects.
I have scraped 4327 objects.
I have scraped 4328 objects.
I have scraped

I have scraped 4579 objects.
I have scraped 4580 objects.
I have scraped 4581 objects.
I have scraped 4582 objects.
I have scraped 4583 objects.
I have scraped 4584 objects.
I have scraped 4585 objects.
I have scraped 4586 objects.
I have scraped 4587 objects.
I have scraped 4588 objects.
I have scraped 4589 objects.
I have scraped 4590 objects.
I have scraped 4591 objects.
I have scraped 4592 objects.
I have scraped 4593 objects.
I have scraped 4594 objects.
I have scraped 4595 objects.
I have scraped 4596 objects.
I have scraped 4597 objects.
I have scraped 4598 objects.
I have scraped 4599 objects.
I have scraped 4600 objects.
I have scraped 4601 objects.
I have scraped 4602 objects.
I have scraped 4603 objects.
I have scraped 4604 objects.
I have scraped 4605 objects.
I have scraped 4606 objects.
I have scraped 4607 objects.
I have scraped 4608 objects.
I have scraped 4609 objects.
I have scraped 4610 objects.
I have scraped 4611 objects.
I have scraped 4612 objects.
I have scraped

I have scraped 4863 objects.
I have scraped 4864 objects.
I have scraped 4865 objects.
I have scraped 4866 objects.
I have scraped 4867 objects.
I have scraped 4868 objects.
I have scraped 4869 objects.
I have scraped 4870 objects.
I have scraped 4871 objects.
I have scraped 4872 objects.
I have scraped 4873 objects.
I have scraped 4874 objects.
I have scraped 4875 objects.
I have scraped 4876 objects.
I have scraped 4877 objects.
I have scraped 4878 objects.
I have scraped 4879 objects.
I have scraped 4880 objects.
I have scraped 4881 objects.
I have scraped 4882 objects.
I have scraped 4883 objects.
I have scraped 4884 objects.
I have scraped 4885 objects.
I have scraped 4886 objects.
I have scraped 4887 objects.
I have scraped 4888 objects.
I have scraped 4889 objects.
I have scraped 4890 objects.
I have scraped 4891 objects.
I have scraped 4892 objects.
I have scraped 4893 objects.
I have scraped 4894 objects.
I have scraped 4895 objects.
I have scraped 4896 objects.
I have scraped

I have scraped 5146 objects.
I have scraped 5147 objects.
I have scraped 5148 objects.
I have scraped 5149 objects.
I have scraped 5150 objects.
I have scraped 5151 objects.
I have scraped 5152 objects.
I have scraped 5153 objects.
I have scraped 5154 objects.
I have scraped 5155 objects.
I have scraped 5156 objects.
I have scraped 5157 objects.
I have scraped 5158 objects.
I have scraped 5159 objects.
I have scraped 5160 objects.
I have scraped 5161 objects.
I have scraped 5162 objects.
I have scraped 5163 objects.
I have scraped 5164 objects.
I have scraped 5165 objects.
I have scraped 5166 objects.
I have scraped 5167 objects.
I have scraped 5168 objects.
I have scraped 5169 objects.
I have scraped 5170 objects.
I have scraped 5171 objects.
I have scraped 5172 objects.
I have scraped 5173 objects.
I have scraped 5174 objects.
I have scraped 5175 objects.
I have scraped 5176 objects.
I have scraped 5177 objects.
I have scraped 5178 objects.
I have scraped 5179 objects.
I have scraped

I have scraped 5429 objects.
I have scraped 5430 objects.
I have scraped 5431 objects.
I have scraped 5432 objects.
I have scraped 5433 objects.
I have scraped 5434 objects.
I have scraped 5435 objects.
I have scraped 5436 objects.
I have scraped 5437 objects.
I have scraped 5438 objects.
I have scraped 5439 objects.
I have scraped 5440 objects.
I have scraped 5441 objects.
I have scraped 5442 objects.
I have scraped 5443 objects.
I have scraped 5444 objects.
I have scraped 5445 objects.
I have scraped 5446 objects.
I have scraped 5447 objects.
I have scraped 5448 objects.
I have scraped 5449 objects.
I have scraped 5450 objects.
I have scraped 5451 objects.
I have scraped 5452 objects.
I have scraped 5453 objects.
I have scraped 5454 objects.
I have scraped 5455 objects.
I have scraped 5456 objects.
I have scraped 5457 objects.
I have scraped 5458 objects.
I have scraped 5459 objects.
I have scraped 5460 objects.
I have scraped 5461 objects.
I have scraped 5462 objects.
I have scraped

I have scraped 5713 objects.
I have scraped 5714 objects.
I have scraped 5715 objects.
I have scraped 5716 objects.
I have scraped 5717 objects.
I have scraped 5718 objects.
I have scraped 5719 objects.
I have scraped 5720 objects.
I have scraped 5721 objects.
I have scraped 5722 objects.
I have scraped 5723 objects.
I have scraped 5724 objects.
I have scraped 5725 objects.
I have scraped 5726 objects.
I have scraped 5727 objects.
I have scraped 5728 objects.
I have scraped 5729 objects.
I have scraped 5730 objects.
I have scraped 5731 objects.
I have scraped 5732 objects.
I have scraped 5733 objects.
I have scraped 5734 objects.
I have scraped 5735 objects.
I have scraped 5736 objects.
I have scraped 5737 objects.
I have scraped 5738 objects.
I have scraped 5739 objects.
I have scraped 5740 objects.
I have scraped 5741 objects.
I have scraped 5742 objects.
I have scraped 5743 objects.
I have scraped 5744 objects.
I have scraped 5745 objects.
I have scraped 5746 objects.
I have scraped

I have scraped 5996 objects.
I have scraped 5997 objects.
I have scraped 5998 objects.
I have scraped 5999 objects.
I have scraped 6000 objects.
I have scraped 6001 objects.
I have scraped 6002 objects.
I have scraped 6003 objects.
I have scraped 6004 objects.
I have scraped 6005 objects.
I have scraped 6006 objects.
I have scraped 6007 objects.
I have scraped 6008 objects.
I have scraped 6009 objects.
I have scraped 6010 objects.
I have scraped 6011 objects.
I have scraped 6012 objects.
I have scraped 6013 objects.
I have scraped 6014 objects.
I have scraped 6015 objects.
I have scraped 6016 objects.
I have scraped 6017 objects.
I have scraped 6018 objects.
I have scraped 6019 objects.
I have scraped 6020 objects.
I have scraped 6021 objects.
I have scraped 6022 objects.
I have scraped 6023 objects.
I have scraped 6024 objects.
I have scraped 6025 objects.
I have scraped 6026 objects.
I have scraped 6027 objects.
I have scraped 6028 objects.
I have scraped 6029 objects.
I have scraped

I have scraped 6279 objects.
I have scraped 6280 objects.
I have scraped 6281 objects.
I have scraped 6282 objects.
I have scraped 6283 objects.
I have scraped 6284 objects.
I have scraped 6285 objects.
I have scraped 6286 objects.
I have scraped 6287 objects.
I have scraped 6288 objects.
I have scraped 6289 objects.
I have scraped 6290 objects.
I have scraped 6291 objects.
I have scraped 6292 objects.
I have scraped 6293 objects.
I have scraped 6294 objects.
I have scraped 6295 objects.
I have scraped 6296 objects.
I have scraped 6297 objects.
I have scraped 6298 objects.
I have scraped 6299 objects.
I have scraped 6300 objects.
I have scraped 6301 objects.
I have scraped 6302 objects.
I have scraped 6303 objects.
I have scraped 6304 objects.
I have scraped 6305 objects.
I have scraped 6306 objects.
I have scraped 6307 objects.
I have scraped 6308 objects.
I have scraped 6309 objects.
I have scraped 6310 objects.
I have scraped 6311 objects.
I have scraped 6312 objects.
I have scraped

I have scraped 6563 objects.
I have scraped 6564 objects.
I have scraped 6565 objects.
I have scraped 6566 objects.
I have scraped 6567 objects.
I have scraped 6568 objects.
I have scraped 6569 objects.
I have scraped 6570 objects.
I have scraped 6571 objects.
I have scraped 6572 objects.
I have scraped 6573 objects.
I have scraped 6574 objects.
I have scraped 6575 objects.
I have scraped 6576 objects.
I have scraped 6577 objects.
I have scraped 6578 objects.
I have scraped 6579 objects.
I have scraped 6580 objects.
I have scraped 6581 objects.
I have scraped 6582 objects.
I have scraped 6583 objects.
I have scraped 6584 objects.
I have scraped 6585 objects.
I have scraped 6586 objects.
I have scraped 6587 objects.
I have scraped 6588 objects.
I have scraped 6589 objects.
I have scraped 6590 objects.
I have scraped 6591 objects.
I have scraped 6592 objects.
I have scraped 6593 objects.
I have scraped 6594 objects.
I have scraped 6595 objects.
I have scraped 6596 objects.
I have scraped

I have scraped 6847 objects.
I have scraped 6848 objects.
I have scraped 6849 objects.
I have scraped 6850 objects.
I have scraped 6851 objects.
I have scraped 6852 objects.
I have scraped 6853 objects.
I have scraped 6854 objects.
I have scraped 6855 objects.
I have scraped 6856 objects.
I have scraped 6857 objects.
I have scraped 6858 objects.
I have scraped 6859 objects.
I have scraped 6860 objects.
I have scraped 6861 objects.
I have scraped 6862 objects.
I have scraped 6863 objects.
I have scraped 6864 objects.
I have scraped 6865 objects.
I have scraped 6866 objects.
I have scraped 6867 objects.
I have scraped 6868 objects.
I have scraped 6869 objects.
I have scraped 6870 objects.
I have scraped 6871 objects.
I have scraped 6872 objects.
I have scraped 6873 objects.
I have scraped 6874 objects.
I have scraped 6875 objects.
I have scraped 6876 objects.
I have scraped 6877 objects.
I have scraped 6878 objects.
I have scraped 6879 objects.
I have scraped 6880 objects.
I have scraped

I have scraped 7131 objects.
I have scraped 7132 objects.
I have scraped 7133 objects.
I have scraped 7134 objects.
I have scraped 7135 objects.
I have scraped 7136 objects.
I have scraped 7137 objects.
I have scraped 7138 objects.
I have scraped 7139 objects.
I have scraped 7140 objects.
I have scraped 7141 objects.
I have scraped 7142 objects.
I have scraped 7143 objects.
I have scraped 7144 objects.
I have scraped 7145 objects.
I have scraped 7146 objects.
I have scraped 7147 objects.
I have scraped 7148 objects.
I have scraped 7149 objects.
I have scraped 7150 objects.
I have scraped 7151 objects.
I have scraped 7152 objects.
I have scraped 7153 objects.
I have scraped 7154 objects.
I have scraped 7155 objects.
I have scraped 7156 objects.
I have scraped 7157 objects.
I have scraped 7158 objects.
I have scraped 7159 objects.
I have scraped 7160 objects.
I have scraped 7161 objects.
I have scraped 7162 objects.
I have scraped 7163 objects.
I have scraped 7164 objects.
I have scraped

### To do
The scraper seems to work, however the entire process could be probably sped up:

* Use better names for variables
* Use asynchronous requests

## Reading the scraped data

In [7]:
import pandas as pd

df = pd.read_json("data.json")

In [8]:
df.shape

(7228, 6)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7228 entries, 0 to 7227
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   fc      7228 non-null   int64 
 1   o       7228 non-null   object
 2   s       7228 non-null   object
 3   prefix  7057 non-null   object
 4   fn      7228 non-null   object
 5   st      7228 non-null   object
dtypes: int64(1), object(5)
memory usage: 338.9+ KB


In [10]:
df.head(30)

fc                                                  o  \
0     404190787                               Comune di Carpanzano   
1     421140039                         Comune di Trarego Viggiona   
2    1447330190  Istituto Superiore di Studi Musicali Claudio M...   
3   83002690275  A.s.v.o. Ambiente Servizi Venezia Orientale S....   
4   80228830586                    Accademia di Belle Arti di Roma   
5   80218690586             Accademia Nazionale di Arte Drammatica   
6    8496131213                                      Acer Campania   
7     139940407                                  Acer Forlì Cesena   
8      80700396                                       Acer Ravenna   
9     493410583                                                ACI   
10    907501001                                                ACI   
11   6363391001                              Agenzia delle Entrate   
12  97254170588                           Agenzia Industrie Difesa   
13   1434200935                            Ambiente Servizi S.p.a.   
14  10811500155                                   Amga Legnano Spa   
15    374200715              Amministrazione Provinciale di Foggia   
16    267390722                               Arca Puglia Centrale   
17   3077790925  Area - Azienda Regionale per L'Edilizia Abitativa   
18   1938950514                                Arezzo Multiservizi   
19   1305930107  Arpal - Agenzia Regionale per La Protezione de...   
20   1474680418                                        Aset S.p.a.   
21   1647800745                                       Asl Brindisi   
22  90062670725                                             Asl Bt   
23   3499370710                                         Asl Foggia   
24  96024110635                                  Asl Napoli 2 Nord   
25    693630550                                   Asm Terni S.p.a.   
26   1210690432  Assm Spa Azienda Specializzata Settore Multise...   
27  95127830164  Autorità di Bacino Lacuale dei Laghi d'Iseo En...   
28    974341000  Autorità di Sistema Portuale del Mar Tirreno C...   
29   6209860482                            Autorità Idrica Toscana   

                                                    s      prefix  \
0   [{'d': '
<p>È ora possibile procedere al pagam...  Comune di    
1   [{'d': '<p>Il servizio di Avviso scadenza cart...  Comune di    
2   [{'d': '<p>Il servizio viene utilizzato dall’e...        None   
3   [{'d': '<p>Il servizio di avvisi di pagamento ...        None   
4   [{'d': '<p>Il servizio viene utilizzato dall’e...        None   
5   [{'d': '<p>Il servizio viene utilizzato dall’e...        None   
6   [{'d': '<p>Il servizio notifica al cittadino, ...        None   
7   [{'d': '<p>In questa sezione potrai trovare gl...        None   
8   [{'d': '<p>In questa sezione potrai trovare gl...        None   
9   [{'d': '<p>I servizi relativi al Bollo Auto of...        None   
10  [{'d': '<p>Il servizio PRA dell’ACI si occupa ...        None   
11  [{'d': '<p>L’incentivo per supportare le famig...        None   
12  [{'d': '<p>Il servizio consente di acquistare ...        None   
13  [{'d': '<p>Il servizio viene utilizzato dall’e...        None   
14  [{'d': '<p>COMUNE DI CANEGRATE - ESPOSIZIONI P...        None   
15  [{'d': '<p>Il servizio comunicazioni istituzio...        None   
16  [{'d': '<p>Il servizio di pagamento locazioni ...        None   
17  [{'d': '<p>Il servizio Alienazione di beni imm...        None   
18  [{'d': '<p>In questa sezione potrai trovare gl...        None   
19  [{'d': '<p>Il servizio permette la consulta de...        None   
20  [{'d': '<p>Il servizio idrico integrato gestit...        None   
21  [{'d': '<p>Il servizio ti invia il promemoria ...        None   
22  [{'d': '<p>Il servizio ti invia un promemoria ...        None   
23  [{'d': '
<p>Il servizio ti invia un promemoria...        None   
24  [{'d': '<p>Il servizio degli avvisi di pagamen...        None   
25  [{'i': '01FK365NWPFMRJEVZHPPHCV1ZX', 'n': 'T

In [99]:
nome_servizio = []
descrizione_servizio = []

for i in df['s']:   
    for s in i:
        nome_servizio.append(s['n'])
        try:
            descrizione_servizio.append(s['d'])
        except:
            descrizione_servizio.append('Nessuna descrizione')


In [102]:
elenco_servizi = list(zip(nome_servizio, descrizione_servizio))

In [105]:
servizi = pd.DataFrame(elenco_servizi)

In [106]:
servizi

0  \
0      Pagamento on line degli affiffi degli alloggi ...   
1      Pagamento on line degli oneri cimiteriali e la...   
2      Pagamento on line dei diritti di segreteria e ...   
3                  Pagamento on line dei passi carrabili   
4                     Pagamento on line dei vari diritti   
...                                                  ...   
88336               Comunicazione Studenti GO-IN Student   
88337                  Comunicazioni Segreteria Studenti   
88338                          Bandi di Gara e Contratti   
88339                                     Libretto Unipi   
88340                      Certificazione verde COVID-19   

                                                       1  
0      \n<p>È ora possibile procedere al pagamento de...  
1      \n<p>È ora possibile procedere al pagamento de...  
2      <p>È ora possibile procedere al pagamento del ...  
3      \n<p>È ora possibile procedere al pagamento de...  
4      \n<p>È ora possibile procedere al pagamento de...  
...                                                  ...  
88336  <p>Servizio di Comunicazione Studenti</p>\n\n<...  
88337  <p>Il servizio consente la ricezione agli Stud...  
88338  <p>Il servizio Bandi di Gara e Contratti offre...  
88339  <p>Il servizio ti avvisa ogni volta che un doc...  
88340  <p>Il servizio consente di ricevere su IO le p...  

[88341 rows x 2 columns]

In [122]:
test = servizi.groupby(0).count().sort_values(by=1, ascending=False)

### Top 50 servizi su App IO

In [128]:
test[0:50]

1
0                                                       
Avviso ritiro documento                             1384
Nuova informazione disponibile dall'Anagrafe        1357
Nuova informazione disponibile dall'Elettorale      1353
Conferma presa in carico istanza web                1350
Informazione su appuntamento uffici demografici     1350
Nuova informazione disponibile dallo Stato civile   1349
Invito ai 17enni stranieri per richiedere citta...  1329
Avviso iscrizione\nomina albi elettorali            1327
Invito a comunitari per iscriversi a liste aggi...  1321
Avviso scadenza carta di identita'                  1306
Avviso scadenza carta d'identità                     838
Servizi alla persona e servizi sociali               724
Politiche educative giovanili e del lavoro           720
Avvocatura civica                                    686
Tributi                                              654
Locazione locali e impianti                          634
Servizi idrici                                       629
Comunicazioni Istituzionali                          627
Ragioneria generale                                  625
Sviluppo economico                                   617
Demografici elettorali e statistici                  599
Patrimonio                                           598
GisMaster Edilizia Privata                           550
Avviso di pagamento                                  492
Anagrafe                                             484
Elettorale                                           469
IMU - Avviso                                         461
Corpo di Polizia Locale                              459
Varie                                                457
Cultura, religioni, pari opportunita', sport         455
Stazione unica appaltante, contratti e gestione...   455
Tutela ambientale, sanita' e igiene pubblica         453
Servizi Cimiteriali                                  444
TARI - Avviso                                        440
GisMaster Cartografia                                433
Sportello online                                     425
Comunicazione apertura pratica cancellazione ir...   385
Comunicazione apertura pratica cancellazione pe...   380
Avviso scadenza permesso di soggiorno                368
Comunicazione stato avanzamento pratiche anagra...   366
Comunicazione da ufficio elettorale                  362
Comunicazione da ufficio anagrafe                    360
GM Pratiche Edilizie                                 356
Tessera elettorale - Avviso                          355
Albo Presidenti di seggio - Avviso                   352
Albo Scrutatori di seggio - Avviso                   350
Prenotazione online appuntamenti                     342
Pagamenti PagoPA                                     342
Titolo di soggiorno - Avviso                         342
Comunicazioni ai cittadini residenti                 339

In [73]:
len(nome_servizio)

88341

In [98]:
#for i, j in df.iterrows():
#    for c in j:

TypeError: 'int' object is not subscriptable

## DA FARE + APPUNTI

To have a better understanding of this data we need to merge it with a database of all public administrations in Italy, using the *fc* column as a **join key**. But let's clean the dataset first.

In [11]:
enti = pd.read_csv('enti.csv')

In [12]:
enti.shape

(22851, 35)

In [13]:
enti.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22851 entries, 0 to 22850
Data columns (total 35 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      22851 non-null  int64  
 1   Codice_IPA               22851 non-null  object 
 2   Denominazione_ente       22851 non-null  object 
 3   Codice_fiscale_ente      22850 non-null  float64
 4   Tipologia                22851 non-null  object 
 5   Codice_Categoria         22851 non-null  object 
 6   Codice_natura            9978 non-null   float64
 7   Codice_ateco             10814 non-null  object 
 8   Ente_in_liquidazione     39 non-null     object 
 9   Codice_MIUR              8310 non-null   object 
 10  Codice_ISTAT             10014 non-null  float64
 11  Acronimo                 4237 non-null   object 
 12  Nome_responsabile        22850 non-null  object 
 13  Cognome_responsabile     22850 non-null  object 
 14  Titolo_responsabile   

In [49]:
enti.head(30)

_id Codice_IPA                                 Denominazione_ente  \
0     1        054           Azienda Unita' Sanitaria Locale Umbria 1   
1     2        055           Azienda Unita' Sanitaria Locale Umbria 2   
2     3        056            Azienda Unita' Sanitaria Locale Viterbo   
3     4        058                    Azienda Sanitaria Locale Roma 5   
4     5   07WB8QXR                                       DOGRE S.R.L.   
5     6        080  Azienda Sanitaria Provinciale N. 5 di Reggio C...   
6     7        092                                    ARNAS G. Brotzu   
7     8        093               Azienda Sanitaria Friuli Occidentale   
8     9        102     Azienda Sanitaria Provinciale di Vibo Valentia   
9    10   15U5YKAC           SOCIETA' ACQUEDOTTO VALLE DEL LAMONE SRL   
10   11   16IODWCU                              ISONTINA AMBIENTE SRL   
11   12      1CMVL           Comunita' Montana Valle del Liri Zona XV   
12   13   1HA36SPJ                    A.S.P. AZIENDA SERVIZI PUBBLICI   
13   14   1MF1PZD3                        Consorzio di bonifica Piave   
14   15   1MPXMKGE  COMMISSARIO DELEGATO STRAORDINARIO EMERGENZA C...   
15   16   1UMEMOPN  Azienda Casa Emilia Romagna della Provincia di...   
16   17   21W263TE  ISTITUTO ISTRUZIONE SUPERIORE "I. E V. FLORIO"...   
17   18   2QR9KMNO            SOCIETA' FERROVIE UDINE CIVIDALE S.R.L.   
18   19   2YNKW2LR                                      Artemedia Srl   
19   20   38HU6HCT               Ordine dei Medici Veterinari Isernia   
20   21   3EYJH0BU    AZIENDA SOCIO SANITARIA LOCALE N. 5 DI ORISTANO   
21   22   3JN4ZMWH      AZIENDA SPECIALE CASA DI RIPOSO DON MOSCHETTA   
22   23   3LM92QZX                              TRENTINO LUNCH S.R.L.   
23   24   3MR8D7NT                                        GORI S.P.A.   
24   25   3S94F5QD                                  ROCCA SERVIZI SRL   
25   26   3T9KJCT0  ISTITUTO COMPRENSIVO VALLE SERINA - SAN PELLEG...   
26   27   3TWJAXOP  SOGGETTO ATTUATORE REGIONE LAZIO EMERGENZA SAN...   
27   28   3VZPQX9H  Azienda Speciale Multiservizi del Comune di Fo...   
28   29   437MRY30            UNIONE DEI COMUNI BASSA VALLE DEL TORTO   
29   30   4EVBLVBD                               ACEGASAPSAMGA S.p.A.   

    Codice_fiscale_ente                    Tipologia Codice_Categoria  \
0          3.301861e+09    Pubbliche Amministrazioni               L7   
1          1.499591e+09    Pubbliche Amministrazioni               L7   
2          1.455571e+09    Pubbliche Amministrazioni               L7   
3          4.733471e+09    Pubbliche Amministrazioni               L7   
4          6.018907e+08  Gestori di Pubblici Servizi              L37   
5          2.638721e+09    Pubbliche Amministrazioni               L7   
6          2.315521e+09    Pubbliche Amministrazioni               L7   
7          1.772891e+09    Pubbliche Amministrazioni               L7   
8          2.866421e+09    Pubbliche Amministrazioni               L7   
9          5.483800e+09  Gestori di Pubblici Servizi              L37   
10         1.123290e+09  Gestori di Pubblici Servizi              L37   
11         8.200441e+10    Pubbliche Amministrazioni              L12   
12         1.213480e+09  Gestori di Pubblici Servizi              L37   
13         4.355020e+09          Stazioni Appaltanti               SA   
14         9.735651e+10    Pubbliche Amministrazioni               L4   
15         8.070040e+07  Gestori di Pubblici Servizi              L37   
16         9.308584e+10    Pubbliche Amministrazioni              L33   
17         2.345670e+09          Stazioni Appaltanti               SA   
18         7.819601e+09  Gestori di Pubblici Servizi              L37   
19         8.005193e+10    Pubbliche Amministrazioni              C14   
20         1.258181e+09    Pubbliche Amministrazioni               L7   
21         4.053410e+09          Stazioni Appaltanti               SA   
22         2.638450e+09          Stazioni Appaltanti         

In [43]:
df['s'][0][0]['d']

'\n<p>È ora possibile procedere al pagamento del servizio.\nPuoi pagare direttamente tramite app IO oppure on line dal sito dell’Ente nella sezione dedicata al PagoPA.</p>\n\n'

In [24]:
nome_servizio = []
descrizione_servizio = []

for i in test:
    nome_servizio.append(i['n'])
    descrizione_servizio.append(i['d'])